Edwardcarlosasigment01-02

edwardcarlosravn opened the pull request:
Exercises solved from assignment 1 and 2

edwardcarlosravn committed:
feat: time-difference solved

edwardcarlosravn committed:
feat: array-rotation solved

edwardcarlosravn committed:
feat: factorial-chain solved

edwardcarlosravn committed:
feat: palindrome-counter solved

edwardcarlosravn committed:
feat: find-majority-element solved

edwardcarlosravn committed:
refactor: another way to solve Exercise01

edwardcarlosravn committed:
feat: 1-callback solved

edwardcarlosravn committed:
feat: 2-promise solved

edwardcarlosravn committed:
feat: 3-async-await solved

edwardcarlosravn committed:
refactor: create getReviewerSubscriptions for better code readability

edwardcarlosravn committed:
feat: create app menu width redline

edwardcarlosravn committed:
refactor: app menu view with promises

edwardcarlosravn committed:
feat: Add a new item finished

edwardcarlosravn committed:
fix(saveItem): add try/catch for error handling

edwardcarlosravn committed:
feat: listAllItems functions created

edwardcarlosravn committed:
refactor: remove var validation for menu iteration

edwardcarlosravn committed:
refactor(id): Change Id generation to auto increment

edwardcarlosravn committed:
feat: updateItemById funtion created

edwardcarlosravn committed:
feat: deleteItemById function created

edwardcarlosravn committed:
feat: showSummary function created

edwardcarlosravn committed:
feat: exportToCSV functions created

edwardcarlosravn committed:
delete: remove color library

juanmdpc commented on 1-js-and-async-programming/1-javascript-fundamentals/1-time-difference.js:
```diff
@@ -16,6 +16,15 @@ timeDifference(7200, 3400); // Expected output: "01:03:20"
+    const differenceInSeconds = Math.abs(a - b);
+    const hours = Math.floor(differenceInSeconds/3600);
+    const minutes = Math.floor(differenceInSeconds%3600/60);
+    const seconds = differenceInSeconds%3600%60;
```
We can use this way to be more concise and clear:
```suggestion
    const hours = Math.floor(differenceInSeconds / 3600);
    const minutes = Math.floor(differenceInSeconds / 60) % 60;
    const seconds = differenceInSeconds % 60;
```

juanmdpc commented on 1-js-and-async-programming/1-javascript-fundamentals/1-time-difference.js:
```diff
@@ -16,6 +16,15 @@ timeDifference(7200, 3400); // Expected output: "01:03:20"
+    const hours = Math.floor(differenceInSeconds/3600);
+    const minutes = Math.floor(differenceInSeconds%3600/60);
+    const seconds = differenceInSeconds%3600%60;
+    const padWidthZero = (num) => (num < 10) ? `0${num}` : String(num);
```
We can leverage the padStart built-in function of a String type
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart
```suggestion
    const padWidthZero = String(num).padStart(2, '0')
```

juanmdpc commented on 1-js-and-async-programming/1-javascript-fundamentals/2-array-rotation.js:
```diff
@@ -17,6 +17,13 @@ rotateArray([1, 2, 3, 4, 5], 7); // Expected output: [3, 4, 5, 1, 2]
+const rotateArray = (arr, n) => {
+    const arrayLength = arr.length;
+    if(arrayLength < 2 || n === 0 || (n % arrayLength) === 0) return arr;
+    n = n % arrayLength;
```
To follow good practices, declare a new variable using the inputs instead of reassigning an input

juanmdpc replied:
Great use of the modulo operator

juanmdpc commented on 1-js-and-async-programming/1-javascript-fundamentals/3-factorial-chain.js:
```diff
@@ -31,6 +31,19 @@ Requirements:
+    let factorial = 1;
+    for(let i = 1; i <= number ; i++){
+        factorial = factorial * i;
+        factorialSum += acc;
```
Did you mean `factorial`?
```suggestion
        factorialSum += factorial;
```

juanmdpc commented on 1-js-and-async-programming/1-javascript-fundamentals/3-factorial-chain.js:
```diff
@@ -31,6 +31,19 @@ Requirements:
+        factorial = factorial * i;
+        factorialSum += acc;
+    };
+    let stringSum = String(factorialSum);
+    while(stringSum.length < lastDigits){
+        stringSum = '0' + stringSum;
+    }
+    return stringSum.slice(-lastDigits);
```
Leverage the built-in functions of a String type
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/padStart
- https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice

```suggestion
    const stringedFactorial = String(factorialSum)
    const mappedFactorial = 
      size >= stringedFactorial.length ? 
        stringedFactorial.padStart(size, '0') :
        stringedFactorial.slice(-size)

   return mappedFactorial
```

edwardcarlosravn replied:
That's actually a better option than manually iterating

juanmdpc commented on 1-js-and-async-programming/1-javascript-fundamentals/4-palindrome-counter.js:
```diff
@@ -31,6 +31,16 @@ Requirements:
 */
 
-const palindromeCounter = (text, minLength) => {};
+const palindromeCounter = (text, minLength) => {
+    const textSplit = text.split(' ');
+    if(textSplit.length < 1 || minLength < 1) return 0;
+    return textSplit.reduce((acc, word) => {
+        const cleanWord = word.toLowerCase().replace(/[^a-z]/g,'');
+        if(cleanWord.length >= minLength && cleanWord === cleanWord.split('').reverse().join('') ){
+            return acc + 1;
+        }
+        return acc;
+    }, 0);
+};
```
Great job!
A suggestion would be to use line breaks or spaces to make the code more readable.

juanmdpc commented on 1-js-and-async-programming/1-javascript-fundamentals/5-find-majority-element.js:
```diff
@@ -19,6 +19,18 @@ findMajorityElement([1, 2, 3, 4, 5]); // Expected output: null
+const findMajorityElement = (arr) => {
+    if(arr.length < 1) return null;
+    if(arr.length === 1) return arr[0];
+    const arraysorter = arr.sort();
```
the sort built-in function modifies the elements of an array in place, and one of the instructions is not to modify the original array.

juanmdpc commented on 1-js-and-async-programming/2-asynchronous-js/1-callback.js:
```diff
@@ -21,7 +21,18 @@
+          console.log(`"All attempts failed."`)
+        }
+      }
+    })
```
We'll get the sixth attempt, but we're on the second attempt instead, and so on.

juanmdpc commented on 1-js-and-async-programming/2-asynchronous-js/3-async-await.js:
```diff
@@ -17,8 +17,40 @@
+      acc[sub] = (acc[sub] || 0) + 1;
+      return acc;
+    },{})
+    return Object.entries(subscriptionCounts).sort((a, b) => b[1] - a[1])[0][0];
```
Make the code readable and easy to understand for everyone
```suggestion
    const [firstSubscription] = Object.entries(subscriptionCounts).sort((a, b) => b[1] - a[1])
    const [subscriptionType] = firstSubscription
    
    return subscriptionType
```

juanmdpc commented on 2-core-nodejs/src/index.js:
great job on dividing the code into different modules

juanmdpc commented on 2-core-nodejs/src/data/storage.js:
```diff
@@ -0,0 +1,90 @@
+        csvContent += row.join(',') + '\n'
+    });
+    try {
+        const filePath = path.join(__dirname, '../../data/wishlistCSV');
```
```suggestion
        const filePath = path.join(__dirname, '../../data/wishlist.csv');
```

edwardcarlosravn committed:
Update 1-js-and-async-programming/1-javascript-fundamentals/1-time-difference.js

Co-authored-by: Juan Pejerrey <40974811+juanmdpc@users.noreply.github.com>

edwardcarlosravn committed:
Update 1-js-and-async-programming/1-javascript-fundamentals/1-time-difference.js

Co-authored-by: Juan Pejerrey <40974811+juanmdpc@users.noreply.github.com>

edwardcarlosravn committed:
Update 1-js-and-async-programming/1-javascript-fundamentals/3-factorial-chain.js

Yes! when I tried to change my variable to a more intuitive one. I forgot to change all of them.

Co-authored-by: Juan Pejerrey <40974811+juanmdpc@users.noreply.github.com>

edwardcarlosravn committed:
Update 1-js-and-async-programming/1-javascript-fundamentals/3-factorial-chain.js

Co-authored-by: Juan Pejerrey <40974811+juanmdpc@users.noreply.github.com>

edwardcarlosravn committed:
Update 1-js-and-async-programming/2-asynchronous-js/3-async-await.js

Co-authored-by: Juan Pejerrey <40974811+juanmdpc@users.noreply.github.com>

edwardcarlosravn committed:
Update 2-core-nodejs/src/data/storage.js

Co-authored-by: Juan Pejerrey <40974811+juanmdpc@users.noreply.github.com>

edwardcarlosravn committed:
refactor: parameter 'n' to 'positionToRotate' for better readability

edwardcarlosravn committed:
refactor: Line breaks to make the code more readable

edwardcarlosravn committed:
refactor: change many variables for better readability

edwardcarlosravn committed:
fix: refactor callback excercise to get correct results
